core/arena/shared_atomic_arena.rs

```diff
@@ -0,0 +1,544 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+use std::alloc::Layout;
+use std::ptr::NonNull;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use parking_lot::lock_api::RawMutex;
+
+use crate::arena::raw_arena::RawArena;
+
+use super::alloc;
+
+#[cfg(debug_assertions)]
+const SIGNATURE: usize = 0x1122334455667788;
+
+pub struct ArenaSharedAtomicReservation<T>(NonNull<ArenaArcData<T>>);
+
+impl<T> Drop for ArenaSharedAtomicReservation<T> {
+  fn drop(&mut self) {
+    panic!("A reservation must be completed or forgotten")
+  }
+}
+
+/// Represents an atomic reference-counted pointer into an arena-allocated object.
+pub struct ArenaArc<T> {
+  ptr: NonNull<ArenaArcData<T>>,
+}
+
+impl<T> ArenaArc<T> {
```

Comment:
Where is shader atomic arena used in this PR?

Reply:
This is used in the OpSet changes that are not part of this PR